README.md
# distribution XML-Compile-WSS

  * My extended documentation: <http://perl.overmeer.net/CPAN/>
  * Development via GitHub: <https://github.com/markov2/perl5-XML-Compile-WSS>
  * Download from CPAN: <ftp://ftp.cpan.org/pub/CPAN/authors/id/M/MA/MARKOV/>
  * Indexed from CPAN: <https://metacpan.org/release/XML-Compile-WSS>

The XML-Compile suite is a large set of modules for various XML related
standards.  This optional component implements the Web Security Standard,
both as stand-alone components in XML as in the header of SOAP messages.

## Development &rarr; Release

Important to know, is that I use an extension on POD to write the manuals.
The "raw" unprocessed version is visible on GitHub.  It will run without
problems, but does not contain manual-pages.

Releases to CPAN are different: "raw" documentation gets removed from
the code and translated into real POD and clean HTML.  This reformatting
is implemented with the OODoc distribution (A name I chose before OpenOffice
existed, sorry for the confusion)

Clone from github for the "raw" version.  For instance, when you want
to contribute a new feature.

On github, you can find the processed version for each release.  But the
better source is CPAN; to get it installed simply run:

```sh
   cpan -i XML::Compile::WSS
```

## Contributing

When you want to contribute to this module, you do not need to provide
a perfect patch... actually: it is nearly impossible to create a patch
which I will merge without modification.  Usually, I need to adapt the
style of code and documentation to my own strict rules.

When you submit an extension, please contribute a set with

1. code

2. code documentation

3. regression tests in t/

**Please note:**
When you contribute in any way, you agree to transfer the copyrights to
Mark Overmeer (you will get the honors in the code and/or ChangeLog).
You also automatically agree that your contribution is released under
the same license as this project: licensed as perl itself.

## Copyright and License

This project is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
See <http://dev.perl.org/licenses/>


ChangeLog

==== version history of XML::Compile:WSS

version 1.15:

	Improvements:
	- move to Git and GitHub.

version 1.14: Mon  8 May 09:15:33 CEST 2017

	Fixes:
	- nonce processing.  rt.cpan.org#121497 [Errietta Kostala]

version 1.13: Tue Jul 21 14:13:35 CEST 2015

	Fixes:
	- try to eliminate confusion about plugins, when more than one
	  ::Schema/::WSDL11 is active in one application.
	  rt.cpan.org#99735 [Jason Adams]

	Improvements:
	- adding wsu_Id does not monopolize the replace hook anymore.

version 1.12: Thu Feb  6 17:12:02 CET 2014

	Fixes:
	- ::SOAP12 is optional, as is the availability of the env12
	  prefix definition. [cpantesters]

	Improvements:
	- change documentation style.

version 1.11: Mon Jan  6 10:39:31 CET 2014

	Fixes:
	- depends on XML::Compile::WSDL11 [cpantesters]

version 1.10: Mon Jan  6 01:07:14 CET 2014

	Improvements:
	- add ::BasicAuth ability to change wsu_Created and wsse_Nonce
	  per operation. [Alex Rutgers]
	- add SOAP 1.2 support

version 1.09: Fri Oct 11 15:52:31 CEST 2013

	Fixes:
	- Accept SOAP-call which passed data as HASH.

version 1.08_02: Sun Sep 22 22:47:26 CEST 2013

	Changes:
	- Move logic from ::SOAP::WSS into ::WSS::Signature

version 1.08_01: Sat Aug 10 00:20:36 CEST 2013

	Changes:
	- calls to ::Signature changed

	Improvements:
	- weaken link to schema
	- use ::Cache::addPrefixes()
	- possible to do only partial plugin preparation: READER and/or WRITER
	- change the way that this actual element signing is triggered, from
	  a two pass message generation into a single.
	- add ::Util WSM10_STR_TRANS, :wsm11 and a few more
	- add ::Util functions wsm_encoded() and wsm_decoded()


version 1.07: Sat Nov 17 12:22:59 CET 2012

	Fixes:
	- remove Digest::SHA1 from the test scripts as wll.
	  [Caleb Cushing]

version 1.06: Thu Nov 15 08:46:30 CET 2012

	Improvements:
	- use Digest::SHA, not ::SHA1.  Add it to Makefile.PL
	  rt.cpan.org#81065 [Chernomyrdin]

version 1.05: Thu Nov  1 15:44:26 CET 2012

	Fixes:
	- flag ::BasicAuth nonce as Base64 encoded.
	  rt.cpan.org#80544 [chernomyrdin]

	Improvements:
	- another nice error when the used initiates ::WSDL before ::WSS

version 1.04: Thu Nov  1 11:44:42 CET 2012

	Changes:
	- split-of ::Signature into separate distribution
	  XML::Compile::WSS::Signature  Also moved a lot of the
	  schema's to that module.

	Fixes:
	- include_namespaces on SignedInfo for Canonicalization.

	Improvements:
	- add mustUnderstand attribute to the security header.

	- remove (not exported) C14N* constants from ::Util, because
	  they are in XML::Compile::C14N::Util

	- permit WSS objects on multiple ::Cache objects [Caleb Cushing]
	  Only one version of WSS per schema.

	- loadSchemas() is now also instance method.

	- clean-up 3 times function _hook_WSU_ID() for one method
	  ::WSS::writerHookWsuId()

	- clean error message when the WSS object is created after the WSDL.

	- ::SOAP::WSS::wssConfigs() renamed to ::features()
	  ::SOAP::WSS::addWSS() renamed to ::addFeature()

version 1.03: Fri Oct 26 17:22:22 CEST 2012

	Fixes:

	- signature: new(remote_pubkey) is documented to accept a filename,
	  but that was not handled. [Karen Etheridge]

	- signature: also include empty prefixList.

	- signature: openssl input format very stringly 64 chars per
	  base64 encoded line.

	Improvements:

	- do not use UNIVERSAL::isa [Karen Etheridge]

	- support for receiving X509v3 certificates

	- signature: remove unused _apply_canon_siginfo()
	  detected by [Karen Etheridge]

version 1.02: Mon Oct 22 11:17:34 CEST 2012

	Fixes:

	- when not using SOAP, then you call create() or check() explicitly.

	- completion of Signature implementation (no real-life testing,
	  but eats its own food, at least)

	Improvements:

	- added regression test for signatures in t/20*

version 1.01: Thu Oct 18 02:49:30 CEST 2012

	Fixes:

	- need for strict order for creation of (at least one) features
	  before compileClient() removed.  Noted by [Cale Cushing]
	  
	Improvements:

	- many documentation improvements.

	- signature: less namespace declarations.

	- removed examples: they are outdated.

version 1.00: Tue Oct 16 16:50:10 CEST 2012

	Changes:

	- deprecated ::WSS::wsseBasicAuth() and ::WSS::wsseTimestamps
	  Still available, tests in t/9?old-*

	Fixes:

	- some export constants for :wsse were incorrect.

	Improvements:

	- option ::WSS::new(wss_version) as alternative to 'version',
	  which is more clear in some circumstances: there are so many
	  versioned namespaces involved!

	- implementation of XML::Compile::WSS::Signature

	- implementation of XML::Compile::WSS::BasicAuth

	- implementation of XML::Compile::WSS::Timestamp

	- promote ::WSS::_dateTime to ::WSS::dateTime()

	- automatically add ValueType="$xsd/dateTime" to timestamps when
	  applicable


version 0.911: Sat Sep  1 13:00:53 CEST 2012

	Fixes:

	- wsseBasicAuth did not use default UTP11_PTEXT anymore
	  [Caleb Cushing]

version 0.91: Fri Aug 31 17:03:22 CEST 2012
	- always permit nonce and created in wsseBasicAuth, not just
	  in combination with UTP11_PDIGEST
	  rt.cpan.org #79315 [David McMath]
	- fix missing xmlns:wsu when wsseBasicAuth it called repetatively.
	  rt.cpan.org #79318 [David McMath]

version 0.90: Thu Aug 30 09:21:56 CEST 2012
	- add support for wsseTimestamp and extended Usertoken, with
	  examples/usertoken/with_help_digest.pl, and t/11digest.t.
	  Contributed by [David McMath]

version 0.14: Tue Jul 31 21:41:25 CEST 2012
	- fix: swap parameters to encode password.  With test patched by
	  rt.cpan.org#78698 [David Tindall Mcmath]

version 0.13: Wed Jan 25 00:10:24 CET 2012
	- add optional parameter for password type to :wsseBasicAuth()
	  [mawasak]

version 0.12: Wed Aug 31 11:31:12 CEST 2011
	- add XML::Compile::WSS::wsseBasicAuth() helper, which was
	  needed by [Caleb Cushing]
	- add examples in examples/usertoken/

version 0.11: Fri Apr 29 11:29:16 CEST 2011
	- add missing dependency to XML::Compile::SOAP [cpantesters]


lib/XML/Compile/WSS.pm
# This code is part of distribution XML-Compile-WSS.  Meta-POD processed
# with OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package XML::Compile::WSS;

use warnings;
use strict;

use Log::Report 'xml-compile-wss';

use XML::Compile::WSS::Util qw/:wss11/;
use XML::Compile::Util      qw/SCHEMA2001/;
use XML::Compile::Schema::BuiltInTypes qw/builtin_type_info/;

use File::Basename          qw/dirname/;
use POSIX                   qw/strftime/;
use Scalar::Util            qw/weaken/;

my %prefixes10 =
  ( ds  => DSIG_NS, wsse => WSSE_10, wsu => WSU_10
  );

my %prefixes11 =
  ( ds  => DSIG_NS, wsse => WSSE_10, wsu => WSU_10
  , wss => WSS_11,  xenc => XENC_NS
  );

my %versions =
  ( '1.0' => { xsddir => 'wss10', prefixes => \%prefixes10 }
  , '1.1' => { xsddir => 'wss11', prefixes => \%prefixes11 }
  );

=chapter NAME
XML::Compile::WSS - OASIS Web Services Security

=chapter SYNOPSIS

 # This modules can be used "stand-alone" ==>
 my $schema = XML::Compile::Cache->new(...);
 my $auth   = XML::Compile::WSS::BasicAuth->new
   (schema => $schema, username => $user, ...);
 my $elem   = $auth->create($doc, $data);

 # ==> or as SOAP client
 my $wss    = XML::Compile::SOAP::WSS->new;
 my $wsdl   = XML::Compile::WSDL11->new($wsdlfn);
 my $auth   = $wss->basicAuth(username => $user, ...);  # once!

 # SOAP call, compile on demand
 my $answer = $wsdl->call($operation, wsse_Security => $auth, %data);
 # same, because "all" defined is default, $auth is in 'all'
 my $answer = $wsdl->call($operation, %data);

 # or SOAP call, explicit compile
 my $call   = $wsdl->compileClient($operation);
 my $answer = $call->(%data);

=chapter DESCRIPTION
The Web Services Security working group of W3C develops a set of
standards which add signatures and encryption to XML.

This module implements features in the C<Security> header.  One header
may contain more than one of these features:
=over 4
=item * timestamps in M<XML::Compile::WSS::Timestamp>
=item * username/password authentication in M<XML::Compile::WSS::BasicAuth>
=item * signing of the body in M<XML::Compile::WSS::Signature>
=item * encryption is not yet supported.  Please hire me to get it implemented.
=back

Furthermore
=over 4
=item * you will certainly need the constants from M<XML::Compile::WSS::Util>.
=item * for SOAP use M<XML::Compile::SOAP::WSS> to create above features.
=back

=chapter METHODS

=section Constructors

=c_method new %options

=requires wss_version '1.1'|MODULE
[1.0] Explicitly state which version WSS needs to be produced.
You may use a version number. You may also use the MODULE
name, which is a namespace constant, provided via C<::Util>.
The only option is currently C<WSS11MODULE>.

=option  version STRING
=default version C<undef>
Alternative for C<wss_version>, but not always as clear.

=option  schema an M<XML::Compile::Cache> object
=default schema C<undef>
Add the WSS extension information to the provided schema.  If not provided
at instantiation, you have to call M<loadSchemas()> before compiling
readers and writers.

=option  prepare 'READER'|'WRITER'|'ALL'|'NONE'
=default prepare 'ALL'
=cut

sub new(@)
{   my $class = shift;
    my $args  = @_==1 ? shift : {@_};
    my $self  = (bless {}, $class)->init($args);
    $self->prepare($args->{prepare} || 'ALL');
    $self;
}

sub init($)
{   my ($self, $args) = @_;
    my $version = $args->{wss_version} || $args->{version}
        or error __x"explicit wss_version required";
    trace "initializing wss $version";

    $version = '1.1'
        if $version eq WSS11MODULE;

    $versions{$version}
        or error __x"unknown wss version {v}, pick from {vs}"
             , v => $version, vs => [keys %versions];
    $self->{XCW_version} = $version;

    if(my $schema = $self->{XCW_schema} = $args->{schema})
    {   weaken $self->{XCW_schema};
        $self->loadSchemas($schema, $version);
    }

    $self;
}

sub prepare($)
{   my ($self, $how) = @_;
    my $schema = $self->schema;

    my ($r, $w)
      = $how eq 'ALL'    ? (1, 1)
      : $how eq 'READER' ? (1, 0)
      : $how eq 'WRITER' ? (0, 1)
      : $how eq 'NONE'   ? (0, 0)
      :                    panic $how;

    $self->prepareWriting($schema) if $w;
    $self->prepareReading($schema) if $r;
    $self;
}
sub prepareWriting($) { $_[0]->{XCW_prepare_w}++; $_[0] }
sub prepareReading($) { $_[0]->{XCW_prepare_r}++; $_[0] }

#-----------
=section Attributes

=method wssVersion
Returns the version number.
=method schema
Returns the schema used to implement this feature.
=cut

sub version()    {shift->{XCW_version}}  # deprecated
sub wssVersion() {shift->{XCW_version}}
sub schema()     {shift->{XCW_schema}}

#-----------
=section Apply

=method create $doc, $security, $data
Adds some WSS element to $security.  The $data is the structure which
is passed to some writer (for instance, the $data which the user
passes to the SOAP call).  There is quite some flexibility in that
structure, so should not be used, in general.
=cut

sub create($$)
{   my $self = shift;
    panic __x"WSS plugin {name} is not prepared for writing", name => ref $self
        unless $self->{XCW_prepare_w};
    $self;
}

=method check $security
Check whether received $security information is correct.  Each active
WSS feature must check whether it finds information for it.
=cut

sub check($)
{   my $self = shift;
    panic __x"WSS plugin {name} is not prepared for reading", name => ref $self
        unless $self->{XCW_prepare_r};
    $self;
}

#-----------
=section Helpers

=method dateTime $time|$string|HASH
Returns a structure which can be used as timestamp, for instance in
C<Created> and C<Expires> fields.  This helper function will help you
use these timestamp fields correctly.

The WSU10 specification defines a free format timestamp.  Of course,
that is very impractical.  Typically a "design by committee" decission.
Also, the standard does not describe the ValueType field, which is often
used to cover this design mistake.
=example

  # Both will get ValueType="$xsd/dateTime"
  Created => time()                 # will get formatted
  Created => '2012-10-14T22:26:21Z' # autodected ValueType

  # Explicit formatting
  Created => { _ => 'this Christmas'
             , ValueType => 'http://per6.org/releasedates'
             };

  # No ValueType added
  Created => '2012-11-01'
=cut

# wsu had "allow anything" date fields, not type dateTime
sub dateTime($)
{   my ($self, $time) = @_;
    return $time if !defined $time || ref $time;

    my $dateTime = builtin_type_info 'dateTime';
    if($time !~ m/[^0-9.]/) { $time = $dateTime->{format}->($time) }
    elsif($dateTime->{check}->($time)) {}
    else {return $time}

     +{ _ => $time
      , ValueType => SCHEMA2001.'/dateTime'
      };
}

#-----------
=section Internals

=ci_method loadSchemas $schema, $version
$schema must extend M<XML::Compile::Cache>.

The $schema settings will may changed a little. For one, the
C<allow_undeclared> flag will be set. Also, C<any_element> will be set to
'ATTEMPT' and C<mixed_elements> to 'STRUCTURAL'.

You can not mix multiple versions of WSS inside one $schema, because
there will be too much confusion about prefixes.
=cut

sub loadSchemas($$)
{   my ($thing, $schema, $version) = @_;
    return if $schema->{XCW_wss_loaded}++;

    $schema->isa('XML::Compile::Cache')
        or error __x"loadSchemas() requires a XML::Compile::Cache object";

    my $def      = $versions{$version};
    my $prefixes = $def->{prefixes};
    $schema->addPrefixes($prefixes);

    my $rewrite = join ',', sort keys %$prefixes;
    $schema->addKeyRewrite("PREFIXED($rewrite)");

    (my $xsddir = __FILE__) =~ s! \.pm$ !/$def->{xsddir}!x;
    my @xsd = glob "$xsddir/*.xsd";

    trace "loading wss schemas $version";

    $schema->importDefinitions
     ( \@xsd

       # Missing from wss-secext-1.1.xsd (schema BUG)  Gladly, all
       # provided schemas have element_form qualified.
     , element_form_default => 'qualified'
     );

    # Another schema bug; attribute wsu:Id not declared qualified
    # Besides, ValueType is often used on timestamps, which are declared
    # as free-format fields (@*!&$#!&^ design committees!)
    my ($wsu10, $xsd) = (WSU_10, SCHEMA2001);
    $schema->importDefinitions( <<__PATCH );
<schema
  xmlns="$xsd"
  xmlns:wsu="$wsu10"
  targetNamespace="$wsu10"
  elementFormDefault="qualified"
  attributeFormDefault="unqualified">
    <attribute name="Id" type="ID" form="qualified" />

    <complexType name="AttributedDateTime">
      <simpleContent>
        <extension base="string">
          <attribute name="ValueType" type="anyURI" />
          <attributeGroup ref="wsu:commonAtts"/>
        </extension>
      </simpleContent>
   </complexType>

</schema>
__PATCH

    $schema->allowUndeclared(1);
    $schema->addCompileOptions('RW'
      , mixed_elements     => 'STRUCTURAL'
      , ignore_unused_tags => qr/^wsu_Id$/
      );

    $schema->anyElement('ATTEMPT');
    $schema;
}

=method writerHookWsuId $type
Creates a hook for an XML producer (writer), to understand wsu:Id on
elements of $type.
=cut

sub writerHookWsuId($)
{   my ($self, $type) = @_;

    my $after = sub
      { my ($doc, $node, $path, $val) = @_;

        my $id = $val->{wsu_Id};
        defined $id or return $node;

        # Some schema explicitly list wsu:Id attributes, we shouldn't add
        # the attribute again.
        if(my $has = $node->getAttributeNS(WSU_10, 'Id')
                  || $node->getAttribute('wsu:Id'))
        {   $has eq $id or warning __x"two wsu:Id attributes: {one} and {two}"
               , one => $id, two => $has;

            return $node;
        }

        $node->setNamespace(WSU_10, 'wsu', 0);
        $node->setAttributeNS(WSU_10, 'Id', $id);
        $node;
      };

     +{ action => 'WRITER', type => $type, after => $after };
}

#---------------------------

=chapter DETAILS

=section Specifications

A huge number of specifications act in this field.  Every self respecting
company has contributed its own implementation into the field.  A lot of
this is B<not supported>, but the list of constants should be complete
in M<XML::Compile::WSS::Util>.

=over 4

=item * XML Security Generic Hybrid Ciphers
F<http://www.w3.org/TR/2011/CR-xmlsec-generic-hybrid-20110303/>, 3 March 2011

=item * XML Signature Properties
F<http://www.w3.org/TR/2011/CR-xmldsig-properties-20110303/>, 3 March 2011

=item * XML Signature Syntax and Processing Version 1.1
F<http://www.w3.org/TR/2011/CR-xmldsig-core1-20110303/>, 3 March 2011

=item * SOAP message security
F<http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0.pdf>, March 2004

=item * XML Signature Syntax and Processing (Second Edition)
F<http://www.w3.org/TR/2008/REC-xmldsig-core-20080610/>, 10 June 2008

=item * RFC4050 Using the ECDSA for XML Digital Signatures
F<http://www.ietf.org/rfc/rfc4050.txt>, april 2005

=item * RFC4051 Additional XML Security Uniform Resource Identifiers (URIs)
F<http://www.ietf.org/rfc/rfc4051.txt>, april 2005

=item * XML Encryption Syntax and Processing
F<http://www.w3.org/TR/2002/REC-xmlenc-core-20021210/>, 10 December 2002

=back
=cut

1;


lib/XML/Compile/WSS/Timestamp.pm
# This code is part of distribution XML-Compile-WSS.  Meta-POD processed
# with OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package XML::Compile::WSS::Timestamp;
use base 'XML::Compile::WSS';

use warnings;
use strict;

use Log::Report  'xml-compile-wss';

use XML::Compile::WSS::Util qw/WSU_10/;

=chapter NAME
XML::Compile::WSS::Timestamp - expiration

=chapter SYNOPSIS

 # used in combination with any XML schema
 my $wss = XML::Compile::WSS::Timestamp->new
   ( ... parametes, some required
   , schema => $anything
   );

 # connects itself to a WSDL
 my $wss  = XML::Compile::SOAP::WSS->new;
 my $wsdl = XML::Compile::WSDL11->new($wsdlfn);
 my $ts   = $wss->timestamp
   ( ... same params, never 'schema'
   );

=chapter DESCRIPTION
The generic Web Service Security protocol is implemented by the super
class M<XML::Compile::WSS>.  This extension implements a timestamp
record.

=chapter METHODS

=section Constructors

=c_method new %options

=default wss_version  '1.1'

=option  created  DATETIME
=default created  C<now>
By default, for each constructed message the current time is taken.
See M<XML::Compile::WSS::dateTime()> for options on DATETIME.  If you
specify an empty string, then the C<Created> node will be skipped.

=option  expires  DATETIME
=default expires  C<undef>
See M<XML::Compile::WSS::dateTime()> for options on DATETIME.  When
not defined, the C<Expires> node will be skipped.

=option  lifetime SECONDS
=default lifetime C<undef>
When C<lifetime> is given and no C<expires>, then the expiration will
be set to the C<created> time plus this lifetime.  In this case, the
created time cannot be specified as formatted DATE.

=option  wsu_Id LABEL
=default wsu_Id undef
Adds a C<wsu:Id> attribute to the C<wsse:Timestamp>.

=cut

sub init($)
{   my ($self, $args) = @_;
    $args->{wss_version} ||= '1.1';
    $self->SUPER::init($args);

    $self->{XCWT_created}  = $args->{created};
    $self->{XCWT_expires}  = $args->{expires};
    $self->{XCWT_lifetime} = $args->{lifetime};
    $self->{XCWT_wsu_id}   = $args->{wsu_Id} || $args->{wsu_id};
    $self;
}

#----------------------------------
=section Attributes

=method created
=method expires
=method lifetime
=method wsuId
=cut

sub created()  {shift->{XCWT_created}}
sub expires()  {shift->{XCWT_expires}}
sub lifetime() {shift->{XCWT_lifetime}}
sub wsuId()    {shift->{XCWT_wsu_id}}

=method timestamps
Returns the "created" and "expires" timestamps.  Both may be undef.
=cut

sub timestamps()
{   my $self    = shift;
    my ($c, $e, $l) = @{$self}{ qw/XCWT_created XCWT_expires XCWT_lifetime/ };
    my ($expires);

    defined $c or $c = time;
    my $created = $c eq '' ? undef : $self->dateTime($c);

    if(!$e && defined $l)
    {    $c !~ m/\D/ or error "lifetime only when created is in seconds";
         $e = $c + $l;
    }
    ($created, $self->dateTime($e));
}

sub prepareWriting($)
{   my ($self, $schema) = @_;
    $self->SUPER::prepareWriting($schema);
    return if $self->{XCWT_stamp};

    my $ts_type = $schema->findName('wsu:Timestamp') ;
    my @hooks   = $self->writerHookWsuId('wsu:TimestampType');
    my $make_ts = $schema->writer($ts_type, include_namespaces => 1
      , hooks => \@hooks);
    $schema->prefixFor(WSU_10);

    $self->{XCWT_stamp} = sub {
        my ($doc, $data) = @_;
        my ($created, $expires) = $self->timestamps;
        $data->{$ts_type} = $make_ts->($doc,
          { wsu_Id      => $self->wsuId
          , wsu_Created => $created
          , wsu_Expires => $expires
          });
        $data;
    };
}

sub create($$)
{   my ($self, $doc, $data) = @_;
    $self->SUPER::create($doc, $data);
    $self->{XCWT_stamp}->($doc, $data);
}

1;


Makefile.PL
# This code is part of distribution XML-Compile-WSS.  Meta-POD processed
# with OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

use ExtUtils::MakeMaker;
use warnings;
use strict;

use 5.010;

# Signatures MUST be upgraded when installed

eval "require XML::Compile::WSS::Signature";
my $version = $XML::Compile::WSS::Signature::VERSION;

if(defined $version && $version lt '2.00')
{   warn <<__WARN;
*************
    You have XML::Compile::WSS::Signature version $version installed,
    but MUST upgrade it to at least version 2.00
*************
__WARN
    sleep 15
}

WriteMakefile
 ( NAME	 		=> 'XML::Compile::WSS'
 , VERSION 		=> '1.15'

 , PREREQ_PM    =>
     { 'XML::Compile'         => '1.50'
     , 'XML::Compile::Cache'  => '1.00'
     , 'XML::Compile::SOAP'   => '3.00'
     , 'XML::Compile::WSDL11' => '3.00'
     , 'Log::Report'          => '0.92'

     , 'Encode'               => 0
     , 'Test::More'           => 0.54
     , 'File::Basename'       => 0
     , 'MIME::Base64'         => 0
     , 'Digest::SHA'          => 0
     , 'Scalar::Util'         => 0
     }
 , EXE_FILES    => []
 , AUTHOR       => 'Mark Overmeer <markov@cpan.org>'
 , ABSTRACT     => 'OASIS Web Services Security'
 , LICENSE      => 'perl_5'

 , META_MERGE =>
      { 'meta-spec' => { version => 2 }
      , resources  =>
          { repository =>
              { type => 'git'
              , url  => 'https://github.com/markov2/perl5-XML-Compile-WSS.git'
              , web  => 'https://github.com/markov2/perl5-XML-Compile-WSS'
              }
          , homepage => 'http://perl.overmeer.net/CPAN/'
          , license  => [ 'http://dev.perl.org/licenses/' ]
          }
      }
 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/xml-compile-wss/raw
DISTDIR         = ../public_html/xml-compile-wss/source

# for POD
FIRST_YEAR      = 2011
EMAIL           = mark@overmeer.net
EXTENDS         = ../XML-Compile:../XML-Compile-Cache:../XML-Compile-C14N:../XML-Compile-WSS-Signature
WEBSITE         = http://perl.overmeer.net/xml-compile/
__POSTAMBLE
